Format commands
==================
The `format` module allows users to convert allele data between different formats. It supports a variety of input and output formats, making it easy to integrate with other tools and workflows.

to_vcf
----------
It will append the genotyped alleles to a existing vcf file.
::

    altools format to_vcf resources/hla_diversity.txt \
        --loci_file resources/gene_locations.tsv \
        --vcf file_to_append_to.vcf

Input file
    Genotype file is tab-separated, where the first column is the sample
    name and pairs of columns for each gene. The header gene name convention is
    "gene" + "gene.1". e.g.::

        "id" "sbgroup" "A" "A.1"
        "sample1" "CEPH" "03:01" "02:01"

.. _loci_file:

--loci_file:
    Additionally, a list of gene locations is required. The file should be
    tab-separated with the following format:
    ::

        gene    start
        HFE    6:26087441
        HLA-A    6:29942554

    The first column is the gene name and the second column is
    (chromosome):(position). This position data can be found in `ensembl`_ or
    `UCSC`_. The sample file used in this repo was obtained from a post in
    `IPD-IMGT`_.

    .. _ensembl: https://www.ensembl.org/index.html
    .. _UCSC: https://genome.ucsc.edu/
    .. _IPD-IMGT: https://www.ebi.ac.uk/ipd/imgt/hla/help/genomics.html

Output file
    *This file must already exist!* because alleles from the input file will be
    appended to it. The file must contain the header of a VCF file and shouldn't
    be compressed.  The header should contain the gene names in the same format
    as the :ref:`loci_file`.  Other alleles can already be present in the file,
    it won't affect the output.
    
    You need ensure that the file contains *ONLY* the samples in the genotype
    file and no more. Otherwise the concatenated alleles won't match the header.
    To filter the samples you can use `bcftools`:
    ::

        cut -d' ' -f1 resources/hla_diversity.txt | tail -n +2 | tr -d '"' | uniq > samples_id.txt
        bcftools view --force-samples -S samples_id.txt test.vcf > filtered.vcf


from_vcf
-----------------------------

Converts from vcf to a genotype table. This command is useful to convert the results
of HLA imputation tools (e.g., `SNP2HLA`_, `HLA-TAPAS`_) to a genotype table that can be
used in other tools.

::

    altools format from_vcf only_hla.vcf --phe input.phe --out output.pyhla

.. _pyhla: https://github.com/felixfan/PyHLA
.. _pypop: https://pypop.org/index.html
.. _SNP2HLA: https://software.broadinstitute.org/mpg/snp2hla/
.. _HLA-TAPAS: https://github.com/immunogenomics/HLA-TAPAS

Input
    A vcf
    The vcf file should be filtered to contain only the HLA genes. You can use
    `bcftools` to do this.
    ::

        bcftools view --include 'ID~"HLA"' raw_imputed.vcf > only_hla.vcf

--phe:
    Additionally, a phenotype file can be provided. This file should follow the
    .phe format of plink files. The first column is the sample ID and the second
    column is the phenotype (0/1 for case/control). The sample IDs must match
    those in the vcf file.

Output
    A .pyhla file compatible with `pyHLA`_ and `PyPop`_ will be generated.


allele_resolution
------------------------------

This script normalizes allele resolutions to a uniform level of the input file,
facilitating association analyses. It ensures that alleles, such as 01 and
01:01, which are essentially identical, are recognized as equal by renaming
them for consistency.
::

    - resolution 1:
        - 01:01 -> 01
        - 01 -> 01
        - 02:03 -> 02
    - resolution 2:
        - 01:01 -> 01:01
        - 01 -> 01:01
        - 02:03 -> 02:03
    - resolution 3:
        - 01:01 -> 01:01:01
        - 01 -> 01:01:01
        - 02:03 -> 02:03:01

If an output file name is not provided, it will be named
`*.[resolution]fields.tsv`, containing the resolved alleles. Up to three
resolution levels are supported (one, two and three).
::

    bash src/alleleTools/refactor/allele_resolution.sh one file1.tsv file2.tsv
    altools format allele_resolution input.alt --resolution 3

Input
    An :ref:`alt_format`.

Output
    An :ref:`alt_format` with normalized allele resolutions.

from_ikmb
------------------------------
Generates a consensus HLA genotype from the result of many HLA genotyping
algorithms. It is tailored to process reports from the `ikmb HLA genotyping`_
pipeline. 
::

    altools format from_ikmb --input "IKMB_Reports/*.json" \
        --output "output.txt" \
        --format pyhla


.. note::
    For more information about the consensus algorithm, see :ref:`consensus_algorithm`.

Input
    Multiple json files follow the format of reports generated by the `ikmb HLA
    genotyping`_ pipeline. These report files should be in a folder called
    `IKMB_Reports/` in a json format.

    .. _ikmb HLA genotyping: https://github.com/ikmb/hla

Output
    An :ref:`alt_format` file will be generated containing the consensus genotype
    for each sample. The default output name is *output.alt*.